### ISSUE TITLE
Update optimizer batch support

### ISSUE THREAD

**Issue by zoq** [mlpack/mlpack#1047]
Right now an optimizer will call the functions:
```
double Evaluate(const arma::mat& coordinates);
void Gradient(const arma::mat& coordinates, arma::mat& gradient);
```
or
```
double Evaluate(const arma::mat& coordinates, const size_t i);
void Gradient(const arma::mat& coordinates, const size_t i, arma::mat& gradient);
```

if it's is a "decomposable" function. For many machine learning algorithms, this abstraction requires the DecomposableFunctionType to hold the data internally, and then use the i parameter, to find the corresponding correct part of the data. Where i is a single part of the decomposable function. This abstraction makes it difficult to integrate batch computation into the existing functions, which could benefit from batch computation. To efficiently implement batch computation, we could call ``Evaluate(coordinates, i, batchSize)`` instead of ``double Evaluate(const arma::mat& coordinates)``, so we can refactor the existing interface to something like this.

```
class DecomposableFunctionType
{
  void Shuffle(); // shuffle any functions
  double Evaluate(const arma::mat& coordinates, const size_t begin, const size_t batchSize);
  double Evaluate(const arma::mat& coordinates);
  void Gradient(const arma::mat& coordinates, const size_t begin, const size_t batchSize, arma::mat& gradient);
  void Gradient(const arma::mat& coordinates, arma::mat& gradient);
  size_t NumFunctions();
};
```

This issue comes out of the discussion for #1034 so there make sure take checkout the discussion there too. We can close this issue when we have found and replaced all instances of the optimizer pattern, and when we have provided good documentation for how to use the updated methods.

**Comment by Rajiv2605** [comment]
@zoq I would like to work on this issue. Please guide me.

**Comment by zoq** [comment]
Sure feel free to work on this issue, I guess a good first step is to go through #1034; let me know if I should clarify anything. Also issue: https://github.com/mlpack/mlpack/issues/1052 might be interesting for you, but don't feel obligated

**Comment by Rajiv2605** [comment]
@zoq I'm feeling a bit lost:sweat_smile: It would be great if you could guide me through this. I mainly want to work on this as I want to contribute something to this library's code and understand how things work around. My previous contribution was a minor one and didn't involve any coding in it(Pull request: #1050). I really like this library as it is coded in C++ which is the fastest language and can get work done faster than the popular Python libraries.

**Comment by zoq** [comment]
Okay, I can show you an example what we have to do, here I refactored the ``SGDTestFunction`` class:

```c++
class SGDTestFunction
{
 public:
  //! Nothing to do for the constructor.
  SGDTestFunction() { }

  //! Return 3 (the number of functions).
  size_t NumFunctions() const { return 3; }

  void Shuffel();
  //! Evaluate a function.
  double Evaluate(const arma::mat& coordinates,
                  const size_t begin,
                  const size_t batchSize) const;

  //! Evaluate the gradient of a function.
  void Gradient(const arma::mat& coordinates,
                const size_t i,
                arma::mat& gradient) const;

  void Gradient(const arma::mat& coordinates,
                const size_t begin,
                const size_t batchSize,
                arma::mat& gradient);
 private:
  arma::Col<size_t> visitationOrder;
};

double SGDTestFunction::Shuffel()
{
  visitationOrder = arma::shuffle(arma::linspace<arma::Col<size_t> >(0,
      (NumFunctions() - 1), NumFunctions()));
}

double SGDTestFunction::Evaluate(
    const arma::mat& coordinates, const size_t begin, const size_t batchSize)
    const
{
  double objective = 0;
  for (size_t i = begin; i < begin + batchSize; ++i)
  {
    switch (visitationOrder(i))
    {
      case 0:
        objective -= std::exp(-std::abs(coordinates[0]));
        break;

      case 1:
        objective += std::pow(coordinates[1], 2);
        break;

      case 2:
        objective += std::pow(coordinates[2], 4) + 3 * std::pow(coordinates[2], 2);
        break;
    }
  }

  return objective;
}

void SGDTestFunction::Gradient(const arma::mat& coordinates,
                               const size_t i,
                               arma::mat& gradient) const
{
  gradient.zeros(3);

  for (size_t i = begin; i < begin + batchSize; ++i)
  {
    switch (visitationOrder(i))
    {
      case 0:
        if (coordinates[0] >= 0)
          gradient[0] = std::exp(-coordinates[0]);
        else
          gradient[0] = -std::exp(coordinates[0]);
        break;

      case 1:
        gradient[1] = 2 * coordinates[1];
        break;

      case 2:
        gradient[2] = 4 * std::pow(coordinates[2], 3) + 6 * coordinates[2];
        break;
    }
  }

  gradient /= batchSize;
}
```

The problem is how we write the Evaluate and Gradient function isn't always that straightforward, since in some cases we can vectorize the expressions to increase the performance.

**Comment by Rajiv2605** [comment]
@zoq So, basically, current optimisers donot support batch optimisation. My role is to give the user control over no. of training samples to be trained at a time which would help in cases like online machine learning that uses stochastic gradient descent. Am I right?

**Comment by zoq** [comment]
right, if we support batches we could see some pretty neat speedups.

**Comment by Rajiv2605** [comment]
@zoq Won't the refractored `Gradient` function code be 
`gradient[0] += std::exp(-coordinates[0]);` 
instead of 
`gradient[0] = std::exp(-coordinates[0]);`

**Comment by Rajiv2605** [comment]
@zoq Why is `visitationOrder` array used?

**Comment by zoq** [comment]
> Won't the refractored Gradient function code be
> gradient[0] += std::exp(-coordinates[0]);
> instead of
> gradient[0] = std::exp(-coordinates[0]);

You are right, I missed that.

> Why is visitationOrder array used?

We decided to introduce a shuffle function, that shuffles the data, in some cases we could just do that, but in this case, there is no data, we can shuffle just some indexes, so I used a vector of indices to shuffle. Does this make sense?

**Comment by Rajiv2605** [comment]
@zoq I have opened a pull request for this issue. It can be found here: #1057

**Comment by rcurtin** [comment]
With #1137 merged this can be closed as fixed.
